contracts/modules/royalty/RoyaltyModule.sol

```diff
@@ -239,30 +296,144 @@ contract RoyaltyModule is
+    function ipRoyaltyVaults(address ipId) external view returns (address) {
+        return _getRoyaltyModuleStorage().ipRoyaltyVaults[ipId];
     }
 
-    /// @notice IERC165 interface support.
+    /// @notice Returns the accumulated royalty policies for a given IP asset
+    /// @param ipId The ID of IP asset
+    function accumulatedRoyaltyPolicies(address ipId) external view returns (address[] memory) {
+        return _getRoyaltyModuleStorage().accumulatedRoyaltyPolicies[ipId].values();
+    }
+
+    /// @notice IERC165 interface support
     function supportsInterface(bytes4 interfaceId) public view virtual override(BaseModule, IERC165) returns (bool) {
         return interfaceId == type(IRoyaltyModule).interfaceId || super.supportsInterface(interfaceId);
     }
 
+    /// @notice Deploys a new ipRoyaltyVault for the given ipId
+    /// @param ipId The ID of IP asset
+    /// @param receiver The address of the receiver
+    /// @return The address of the deployed ipRoyaltyVault
+    function _deployIpRoyaltyVault(address ipId, address receiver) internal returns (address) {
+        RoyaltyModuleStorage storage $ = _getRoyaltyModuleStorage();
+
+        address ipRoyaltyVault = address(new BeaconProxy(ipRoyaltyVaultBeacon(), ""));
+        IIpRoyaltyVault(ipRoyaltyVault).initialize("Royalty Token", "RT", TOTAL_RT_SUPPLY, ipId, receiver);
+        $.ipRoyaltyVaults[ipId] = ipRoyaltyVault;
+
+        return ipRoyaltyVault;
+    }
+
+    /// @notice Distributes royalty tokens to the royalty policies of the ancestors IP assets
+    /// @param ipId The ID of the IP asset
+    /// @param parentIpIds The parent IP assets
+    /// @param licenseRoyaltyPolicies The royalty policies of the each parent license
+    /// @param licensesPercent The license percentage of the licenses being minted
+    /// @param ipRoyaltyVault The address of the ipRoyaltyVault
+    function _distributeRoyaltyTokensToPolicies(
+        address ipId,
+        address[] calldata parentIpIds,
+        address[] calldata licenseRoyaltyPolicies,
+        uint32[] calldata licensesPercent,
+        address ipRoyaltyVault
+    ) internal {
+        RoyaltyModuleStorage storage $ = _getRoyaltyModuleStorage();
+
+        uint32 totalRtsRequiredToLink;
+        for (uint256 i = 0; i < parentIpIds.length; i++) {
+            if (parentIpIds[i] == address(0)) revert Errors.RoyaltyModule__ZeroParentIpId();
+            if (licenseRoyaltyPolicies[i] == address(0)) revert Errors.RoyaltyModule__ZeroRoyaltyPolicy();
+            _addToAccumulatedRoyaltyPolicies(parentIpIds[i], licenseRoyaltyPolicies[i]);
+            address[] memory accParentRoyaltyPolicies = $.accumulatedRoyaltyPolicies[parentIpIds[i]].values();
+            for (uint256 j = 0; j < accParentRoyaltyPolicies.length; j++) {
+                // add the parent ancestor royalty policies to the child
+                _addToAccumulatedRoyaltyPolicies(ipId, accParentRoyaltyPolicies[j]);
```

Comment:
Would there be scalability issue when a parent IP has a large amount of children? e.g. a parent has more than 10K children.

Reply:
There can only be 8 parents at the moment - so that is the limit of the loop.

Reply:
The scalability issue is more related to how many royalty policies exist in an IP ancestry (no issue with the number of children). It is a boundless loop technically - in practice it will depend on how many royalty policies the ancestors have - which at the moment we don't expect to be that many.

Reply:
If we really want to be on the safe side - we can put a cap on the total amount on different policies allowed in an IP ancestry.

Reply:
Addressed in: https://github.com/storyprotocol/protocol-core-v1/pull/203/commits/03bee8db346021744c385bd18d02719b1ae2d2ae